File: packages/agent/src/quarantine.ts
import { existsSync } from "node:fs";
import { mkdir, rename } from "node:fs/promises";
import path from "node:path";
import { config } from "./config.js";
import { fetchQuarantineCommands, completeQuarantineCommand } from "./client.js";
import { computeQuarantinePath, QUARANTINE_DIR_NAME } from "./quarantinePath.js";

/**
 * Whatever "move this file into quarantine" means for the active source —
 * local mode does it with `node:fs` directly (below); `SmbSource.quarantine`
 * (sources/smb.ts) is the other implementation. M365 and Google Drive don't
 * have one — those connectors never call `startQuarantinePolling` at all
 * (see index.ts), since the classification worker never suggests
 * FILE_QUARANTINE for them in the first place (watchedRoot.ts).
 */
export interface QuarantineTarget {
  quarantine(filePath: string): Promise<string>;
}

async function quarantineLocalFile(filePath: string): Promise<string> {
  const quarantineDir = path.join(config.watchPath, QUARANTINE_DIR_NAME);
  await mkdir(quarantineDir, { recursive: true });

  const existingNames = new Set<string>();
  let destination = computeQuarantinePath(config.watchPath, filePath);
  while (existsSync(destination)) {
    existingNames.add(path.basename(destination));
    destination = computeQuarantinePath(config.watchPath, filePath, existingNames);
  }

  await rename(filePath, destination);
  return destination;
}

export const localQuarantineTarget: QuarantineTarget = { quarantine: quarantineLocalFile };

async function quarantineOne(target: QuarantineTarget, filePath: string): Promise<{ ok: true; destination: string } | { ok: false; error: string }> {
  try {
    const destination = await target.quarantine(filePath);
    return { ok: true, destination };
  } catch (err) {
    const error = (err as NodeJS.ErrnoException).code === "ENOENT"
      ? "file no longer exists"
      : (err as Error).message;
    return { ok: false, error };
  }
}

// A RANSOMWARE_RATE command can carry many paths (see agentCommands.ts) but
// the ResponseAction it's attached to has one status/resultMessage, not one
// per file — approving it is one click, not N. So this is deliberately
// all-or-nothing at the status level: EXECUTED only if every path
// succeeded, FAILED otherwise, with per-file detail folded into the text
// message rather than the schema. A partially-succeeded burst quarantine
// is visible in the dashboard's message text, not as its own status — a
// known simplification, not an oversight.
async function pollOnce(target: QuarantineTarget): Promise<void> {
  const commands = await fetchQuarantineCommands();
  for (const command of commands) {
    const outcomes = await Promise.all(
      command.paths.map(async (path) => ({ path, result: await quarantineOne(target, path) })),
    );
    const succeeded = outcomes.filter((o): o is { path: string; result: { ok: true; destination: string } } => o.result.ok);
    const failed = outcomes.filter((o): o is { path: string; result: { ok: false; error: string } } => !o.result.ok);

    for (const { path, result } of outcomes) {
      if (result.ok) console.log(`quarantined ${path} -> ${result.destination}`);
      else console.error(`failed to quarantine ${path}: ${result.error}`);
    }

    const message =
      command.paths.length === 1
        ? succeeded.length === 1
          ? `moved to ${succeeded[0].result.destination}`
          : `failed to quarantine ${failed[0].path}: ${failed[0].result.error}`
        : `quarantined ${succeeded.length}/${command.paths.length} file(s)${
            failed.length > 0 ? `; failed: ${failed.map((f) => `${f.path} (${f.result.error})`).join(", ")}` : ""
          }`;

    await completeQuarantineCommand(command.id, failed.length === 0, message);
  }
}

export function startQuarantinePolling(intervalMs: number, target: QuarantineTarget = localQuarantineTarget): void {
  console.log(`polling for quarantine commands every ${intervalMs}ms`);
  pollOnce(target).catch((err) => console.error("quarantine poll failed", err));
  setInterval(() => {
    pollOnce(target).catch((err) => console.error("quarantine poll failed", err));
  }, intervalMs);
}

File: native-agent/cmd/agent/quarantine.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/logikos-dsp/native-agent/internal/client"
	"github.com/logikos-dsp/native-agent/internal/config"
)

const quarantineDirName = ".logikos-quarantine" // must match packages/agent/src/quarantinePath.ts's QUARANTINE_DIR_NAME

// computeQuarantinePath matches packages/agent/src/quarantinePath.ts's
// computeQuarantinePath exactly — same collision-avoidance naming
// (` (1)`, ` (2)`, ...), same signature shape (watchedRoot + full path +
// already-seen names for this call), so quarantined filenames look
// identical regardless of which agent implementation did the moving.
func computeQuarantinePath(watchedRoot, filePath string, existingNames map[string]bool) string {
	quarantineDir := filepath.Join(watchedRoot, quarantineDirName)
	ext := filepath.Ext(filePath)
	base := strings.TrimSuffix(filepath.Base(filePath), ext)

	candidate := base + ext
	suffix := 1
	for existingNames[candidate] {
		candidate = fmt.Sprintf("%s (%d)%s", base, suffix, ext)
		suffix++
	}
	return filepath.Join(quarantineDir, candidate)
}

func quarantineOne(watchPath, filePath string) (destination string, err error) {
	quarantineDir := filepath.Join(watchPath, quarantineDirName)
	if err := os.MkdirAll(quarantineDir, 0o755); err != nil {
		return "", err
	}

	existingNames := map[string]bool{}
	destination = computeQuarantinePath(watchPath, filePath, existingNames)
	for {
		if _, statErr := os.Stat(destination); errors.Is(statErr, os.ErrNotExist) {
			break
		}
		existingNames[filepath.Base(destination)] = true
		destination = computeQuarantinePath(watchPath, filePath, existingNames)
	}

	if err := os.Rename(filePath, destination); err != nil {
		return "", err
	}
	return destination, nil
}

// runQuarantinePolling mirrors packages/agent/src/quarantine.ts's
// pollOnce/startQuarantinePolling exactly, including the all-or-nothing
// status for a multi-path (ransomware-burst) command — one ResponseAction
// has one status, not one per file, so a command with several paths
// reports EXECUTED only if every one succeeded, FAILED otherwise with
// per-file detail folded into the text message. See ARCHITECTURE.md's
// "Ransomware-burst quarantine" note for why that tradeoff was accepted
// rather than changing the schema.
func runQuarantinePolling(c *client.Client, cfg config.Config) {
	pollOnce := func() {
		commands, err := c.FetchQuarantineCommands(cfg.AgentKey)
		if err != nil {
			return // transport error — skip this tick quietly, next poll interval retries
		}
		for _, cmd := range commands {
			type outcome struct {
				path        string
				destination string
				err         error
			}
			outcomes := make([]outcome, len(cmd.Paths))
			for i, p := range cmd.Paths {
				dest, err := quarantineOne(cfg.WatchPath, p)
				outcomes[i] = outcome{path: p, destination: dest, err: err}
			}

			var failed []outcome
			var succeeded []outcome
			for _, o := range outcomes {
				if o.err != nil {
					failed = append(failed, o)
				} else {
					succeeded = append(succeeded, o)
				}
			}

			var message string
			if len(cmd.Paths) == 1 {
				if len(succeeded) == 1 {
					message = fmt.Sprintf("moved to %s", succeeded[0].destination)
				} else {
					message = fmt.Sprintf("failed to quarantine %s: %v", failed[0].path, failed[0].err)
				}
			} else {
				message = fmt.Sprintf("quarantined %d/%d file(s)", len(succeeded), len(cmd.Paths))
				if len(failed) > 0 {
					parts := make([]string, len(failed))
					for i, f := range failed {
						parts[i] = fmt.Sprintf("%s (%v)", f.path, f.err)
					}
					message += "; failed: " + strings.Join(parts, ", ")
				}
			}

			if err := c.CompleteQuarantineCommand(cfg.AgentKey, cmd.ID, len(failed) == 0, message); err != nil {
				continue // next poll picks the command back up if it's still APPROVED — same recovery path as the TS agent
			}
		}
	}

	ticker := time.NewTicker(time.Duration(cfg.QuarantinePollIntervalMs) * time.Millisecond)
	defer ticker.Stop()
	pollOnce()
	for range ticker.C {
		pollOnce()
	}
}
